File: nodes/motor_controller.py
#!/usr/bin/env python3
"""
Motor controller node — open loop.
Subscribes to /cmd_vel, drives TB6612 via GPIO.
"""

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
import lgpio
from nav_msgs.msg import Odometry
from std_msgs.msg import Header
import math
import gpiod
import threading
import datetime


# --- Pin definitions (BCM numbering) ---
AIN1 = 27  # Motor A (Right) direction
AIN2 = 17  # Motor A (Right) direction
PWMA = 13  # Motor A (Right) speed

BIN1 = 26  # Motor B (Left) direction
BIN2 = 19  # Motor B (Left) direction
PWMB = 12  # Motor B (Left) speed

STBY = None  # Pulled HIGH via resistor — no GPIO control needed

# --- Robot physical parameters ---
WHEEL_BASE = 0.34       # metres — distance between wheels, measure and update
MAX_SPEED = 100         # PWM duty cycle max (0-100)

MOTOR_PINS = [AIN1, AIN2, PWMA, BIN1, BIN2, PWMB]

#PID controller tunable parameters
pid_kp = 0.01
pid_ki = 0.005
pid_kd = 0
windup_limit = 50




class MotorController(Node):

    def __init__(self):
    
        super().__init__('motor_controller')
        self.get_logger().info('Motor controller starting...')

        # Init GPIO
        self.chip = lgpio.gpiochip_open(4)
        for pin in MOTOR_PINS:
            lgpio.gpio_claim_output(self.chip, pin, 0)

        # PWM setup — 1000 Hz
        lgpio.tx_pwm(self.chip, PWMA, 1000, 0)
        lgpio.tx_pwm(self.chip, PWMB, 1000, 0)

        # Encoder setup via gpiod kernel interrupts
        self.enc_a_count = 0
        self.enc_b_count = 0
        self._enc_lock = threading.Lock()

        self.gpio_chip = gpiod.Chip('/dev/gpiochip4')

        self.enc_a_request = self.gpio_chip.request_lines(
            consumer='enc_a',
            config={21: gpiod.LineSettings(edge_detection=gpiod.line.Edge.RISING, debounce_period=datetime.timedelta(microseconds=0.5))}
        )
        self.enc_b_request = self.gpio_chip.request_lines(
            consumer='enc_b',
            config={24: gpiod.LineSettings(edge_detection=gpiod.line.Edge.RISING, debounce_period=datetime.timedelta(microseconds=0.5))}
        )

        self.enc_a_thread = threading.Thread(target=self._watch_enc_a, daemon=True)
        self.enc_b_thread = threading.Thread(target=self._watch_enc_b, daemon=True)
        self.enc_a_thread.start()
        self.enc_b_thread.start()

        # ROS subscriber
        self.sub = self.create_subscription(
            Twist,
            '/cmd_vel',
            self.cmd_vel_callback,
            10
        )
        

        self.get_logger().info('Motor controller ready.')

        self.odom_pub = self.create_publisher(Odometry, '/odom', 10)
        self.x = 0.0
        self.y = 0.0
        self.theta = 0.0
        self.last_enc_a = 0
        self.last_enc_b = 0
        self.create_timer(0.02, self.publish_odom)  # 50Hz

        # PID controller
        self.pid_integral = 0.0
        self.pid_prev_error = 0.0
        self.correction = 0.0
        self.linear = 0.0
        self.angular = 0.0
        self.debug_counter = 0

        self.active = False

    def cmd_vel_callback(self, msg):
        self.get_logger().info(f'cmd_vel: linear={msg.linear.x:.2f} angular={msg.angular.z:.2f}')
        self.get_logger().info(f'enc_a={self.enc_a_count} enc_b={self.enc_b_count}')
        self.linear = msg.linear.x
        self.angular = msg.angular.z
        self.active = True

    def drive_motor_a(self, speed):
        duty = min(abs(speed) * MAX_SPEED, MAX_SPEED)
        if speed > 0:
            lgpio.gpio_write(self.chip, AIN1, 1) # was 0
            lgpio.gpio_write(self.chip, AIN2, 0) # was 1
        elif speed < 0:
            lgpio.gpio_write(self.chip, AIN1, 0) # was 1
            lgpio.gpio_write(self.chip, AIN2, 1) # was 0
        else:
            lgpio.gpio_write(self.chip, AIN1, 0)
            lgpio.gpio_write(self.chip, AIN2, 0)
        lgpio.tx_pwm(self.chip, PWMA, 1000, duty)

    def drive_motor_b(self, speed):
        duty = min(abs(speed) * MAX_SPEED, MAX_SPEED)
        if speed > 0:
            lgpio.gpio_write(self.chip, BIN1, 0)
            lgpio.gpio_write(self.chip, BIN2, 1)
        elif speed < 0:
            lgpio.gpio_write(self.chip, BIN1, 1)
            lgpio.gpio_write(self.chip, BIN2, 0)
        else:
            lgpio.gpio_write(self.chip, BIN1, 0)
            lgpio.gpio_write(self.chip, BIN2, 0)
        lgpio.tx_pwm(self.chip, PWMB, 1000, duty)

    def stop_motors(self):
        lgpio.gpio_write(self.chip, AIN1, 0)
        lgpio.gpio_write(self.chip, AIN2, 0)
        lgpio.gpio_write(self.chip, BIN1, 0)
        lgpio.gpio_write(self.chip, BIN2, 0)
        lgpio.tx_pwm(self.chip, PWMA, 1000, 0)
        lgpio.tx_pwm(self.chip, PWMB, 1000, 0)

    def destroy_node(self):
        self.stop_motors()
        lgpio.gpiochip_close(self.chip)
        self.enc_a_request.release()
        self.enc_b_request.release()
        self.gpio_chip.close()
        super().destroy_node()


    def publish_odom(self):
        PULSES_PER_REV = 1050
        WHEEL_CIRCUMFERENCE = 0.1382

        delta_a = self.enc_a_count - self.last_enc_a
        delta_b = self.enc_b_count - self.last_enc_b
        self.last_enc_a = self.enc_a_count
        self.last_enc_b = self.enc_b_count

        dist_a = (delta_a / PULSES_PER_REV) * WHEEL_CIRCUMFERENCE
        dist_b = (delta_b / PULSES_PER_REV) * WHEEL_CIRCUMFERENCE

        error = delta_a - delta_b
        self.pid_integral += error * 0.02
        self.pid_integral = max(-windup_limit, min(windup_limit, self.pid_integral))
        derivative = (error - self.pid_prev_error) / 0.02
        self.correction = pid_kp * error + pid_ki * self.pid_integral + pid_kd * derivative
        self.debug_counter += 1
        if self.debug_counter % 25 == 0:
            self.get_logger().info(f'correction={self.correction:.4f} integral={self.pid_integral:.4f} error={error}')
        self.pid_prev_error = error

        right = self.linear - (self.angular * WHEEL_BASE / 2.0) - self.correction
        left = self.linear + (self.angular * WHEEL_BASE / 2.0) + self.correction

        if self.active and self.linear != 0.0:
            self.drive_motor_a(right)
            self.drive_motor_b(left)
        else:
            self.stop_motors()

        dist = (dist_a + dist_b) / 2.0
        dtheta = (dist_a - dist_b) / WHEEL_BASE
        self.x += dist * math.cos(self.theta)
        self.y += dist * math.sin(self.theta)
        self.theta += dtheta

        msg = Odometry()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = 'odom'
        msg.child_frame_id = 'base_link'
        msg.pose.pose.position.x = self.x
        msg.pose.pose.position.y = self.y
        msg.twist.twist.linear.x = dist / 0.02
        msg.twist.twist.angular.z = dtheta / 0.02
        self.odom_pub.publish(msg)
    
    def _watch_enc_a(self):
        while rclpy.ok():
            if self.enc_a_request.wait_edge_events(datetime.timedelta(seconds=1)):
                self.enc_a_request.read_edge_events()
                with self._enc_lock:
                    self.enc_a_count += 1

    def _watch_enc_b(self):
        while rclpy.ok():
            if self.enc_b_request.wait_edge_events(datetime.timedelta(seconds=1)):
                self.enc_b_request.read_edge_events()
                with self._enc_lock:
                    self.enc_b_count += 1


def main(args=None):
    rclpy.init(args=args)
    node = MotorController()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: rover_control/src/motor_controller.cpp
// motor_controller.cpp
// TB6612 differential drive — ROS2 Jazzy — Raspberry Pi

#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <lgpio.h>
#include <gpiod.h>
#include <thread>
#include <mutex>
#include <atomic>
#include <cmath>
#include <algorithm>
#include <chrono>

// ─── PIN DEFINITIONS (BCM) ────────────────────────────────────────────────────
constexpr int AIN1 = 27;
constexpr int AIN2 = 17;
constexpr int PWMA = 13;

constexpr int BIN1 = 26;
constexpr int BIN2 = 19;
constexpr int PWMB = 12;

// ─── ROBOT PARAMETERS ─────────────────────────────────────────────────────────
constexpr double WHEEL_BASE          = 0.34;
constexpr double MAX_SPEED           = 100.0;
constexpr int    PULSES_PER_REV      = 1050;
constexpr double WHEEL_CIRCUMFERENCE = 0.1382;
constexpr double DT                  = 0.02;   // 50 Hz timer period (seconds)

// ─── PID TUNING ───────────────────────────────────────────────────────────────
// TODO: set these once you have written and tested the PID
constexpr double PID_KP       = 0.001;
constexpr double PID_KI       = 0.0;
constexpr double PID_KD       = 0.0;
constexpr double WINDUP_LIMIT = 50.0;
constexpr double MOTOR_B_TRIM = 0.971;


// ─── PID STATE ────────────────────────────────────────────────────────────────
// These three variables are all the memory the PID needs between timer ticks.
// You own these — do not touch them anywhere except inside computePID().
struct PidState 
{
    double integral   = 0.0;
    double prev_error = 0.0;
};

PidState pid_state_a_;
PidState pid_state_b_;

double computePID(double error, PidState& state)
{

double p = PID_KP * error;
state.integral += PID_KI * error * DT;
state.integral = std::clamp(state.integral, -WINDUP_LIMIT, WINDUP_LIMIT);
double d = PID_KD * (error - state.prev_error) / DT;
state.prev_error = error;
return p + state.integral + d;

}



// ─── MOTOR CONTROLLER NODE ────────────────────────────────────────────────────
class MotorController : public rclcpp::Node
{
public:
    MotorController() : Node("motor_controller")
    {
        RCLCPP_INFO(get_logger(), "Motor controller starting...");

        initGPIO();
        initEncoders();

        cmd_vel_sub_ = create_subscription<geometry_msgs::msg::Twist>(
            "/cmd_vel", 10,
            [this](const geometry_msgs::msg::Twist::SharedPtr msg) {
                cmdVelCallback(msg);
            });

        odom_pub_ = create_publisher<nav_msgs::msg::Odometry>("/odom", 10);

        timer_ = create_wall_timer(
            std::chrono::milliseconds(20),
            [this]() { timerCallback(); });

        RCLCPP_INFO(get_logger(), "Motor controller ready.");
    }

    ~MotorController()
    {
        stopMotors();
        lgGpiochipClose(chip_);
        gpiod_line_release(enc_a_line_);
        gpiod_line_release(enc_b_line_);
        gpiod_chip_close(gpiod_chip_);
    }

private:

    // ── GPIO / HARDWARE ───────────────────────────────────────────────────────
    int chip_;

    void initGPIO()
    {
        chip_ = lgGpiochipOpen(4);
        for (int pin : {AIN1, AIN2, PWMA, BIN1, BIN2, PWMB})
            lgGpioClaimOutput(chip_, 0, pin, 0);
        lgTxPwm(chip_, PWMA, 1000, 0, 0, 0);
        lgTxPwm(chip_, PWMB, 1000, 0, 0, 0);
    }

    void driveMotorA(double speed)
    {
        double duty = std::min(std::abs(speed) * MAX_SPEED, MAX_SPEED);
        lgGpioWrite(chip_, AIN1, speed > 0 ? 1 : 0);
        lgGpioWrite(chip_, AIN2, speed > 0 ? 0 : 1);
        if (speed == 0.0) { lgGpioWrite(chip_, AIN1, 0); lgGpioWrite(chip_, AIN2, 0); }
        lgTxPwm(chip_, PWMA, 1000, duty, 0, 0);       
    }

    void driveMotorB(double speed)
    {
        double duty = std::min(std::abs(speed) * MAX_SPEED, MAX_SPEED);
        lgGpioWrite(chip_, BIN1, speed > 0 ? 0 : 1);
        lgGpioWrite(chip_, BIN2, speed > 0 ? 1 : 0);
        if (speed == 0.0) { lgGpioWrite(chip_, BIN1, 0); lgGpioWrite(chip_, BIN2, 0); }
        lgTxPwm(chip_, PWMB, 1000, duty, 0, 0);
    }

    void stopMotors()
    {
        lgGpioWrite(chip_, AIN1, 0); lgGpioWrite(chip_, AIN2, 0);
        lgGpioWrite(chip_, BIN1, 0); lgGpioWrite(chip_, BIN2, 0);
        lgTxPwm(chip_, PWMA, 1000, 0, 0, 0);
        lgTxPwm(chip_, PWMB, 1000, 0, 0, 0);
    }

    // ── ENCODERS ──────────────────────────────────────────────────────────────
    std::atomic<int> enc_a_count_{0};
    std::atomic<int> enc_b_count_{0};
    std::mutex       enc_lock_;

    struct gpiod_chip* gpiod_chip_;
    struct gpiod_line* enc_a_line_;
    struct gpiod_line* enc_b_line_;
    std::thread enc_a_thread_;
    std::thread enc_b_thread_;

    void initEncoders()
    {
        gpiod_chip_ = gpiod_chip_open("/dev/gpiochip4");
        enc_a_line_ = gpiod_chip_get_line(gpiod_chip_, 21);
        enc_b_line_ = gpiod_chip_get_line(gpiod_chip_, 24);
        gpiod_line_request_rising_edge_events(enc_a_line_, "enc_a");
        gpiod_line_request_rising_edge_events(enc_b_line_, "enc_b");
        enc_a_thread_ = std::thread([this]() { watchEncoder(enc_a_line_, enc_a_count_); });
        enc_b_thread_ = std::thread([this]() { watchEncoder(enc_b_line_, enc_b_count_); });
        enc_a_thread_.detach();
        enc_b_thread_.detach();
    }

    void watchEncoder(struct gpiod_line* line, std::atomic<int>& count)
    {
        struct gpiod_line_event event;
        struct timespec timeout = {1, 0};
        while (rclcpp::ok()) {
            int ret = gpiod_line_event_wait(line, &timeout);
            if (ret == 1) {
                gpiod_line_event_read(line, &event);
                count++;
                

            }
        }
    }

    // ── CMD_VEL ───────────────────────────────────────────────────────────────
    double linear_  = 0.0;
    double angular_ = 0.0;
    bool   active_  = false;

    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_sub_;

    void cmdVelCallback(const geometry_msgs::msg::Twist::SharedPtr msg)
    {
        linear_  = msg->linear.x;
        //angular_ = msg->angular.z;
        angular_ = 0.0;
        active_  = true;
    }

    // ── ODOMETRY ──────────────────────────────────────────────────────────────
    double x_     = 0.0;
    double y_     = 0.0;
    double theta_ = 0.0;
    int    last_enc_a_ = 0;
    int    last_enc_b_ = 0;
    int    debug_counter_ = 0;

    rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;

    // ── PID STATE INSTANCE ────────────────────────────────────────────────────
    PidState pid_state_;

    // ── MAIN TIMER CALLBACK ───────────────────────────────────────────────────
    rclcpp::TimerBase::SharedPtr timer_;

    void timerCallback()
    {
        int current_a = enc_a_count_.load();
        int current_b = enc_b_count_.load();
        int delta_a = current_a - last_enc_a_;
        int delta_b = current_b - last_enc_b_;
        last_enc_a_ = current_a;
        last_enc_b_ = current_b;

        // ── PID ───────────────────────────────────────────────────────────────
        double target_a = linear_ - (angular_ * WHEEL_BASE / 2.0);
        double target_b = (linear_ + (angular_ * WHEEL_BASE / 2.0)) * MOTOR_B_TRIM;
        double target_counts_a = target_a * (PULSES_PER_REV * DT / WHEEL_CIRCUMFERENCE);
        double target_counts_b = target_b * (PULSES_PER_REV * DT / WHEEL_CIRCUMFERENCE);
        double error_a = target_counts_a - delta_a;
        double error_b = target_counts_b - delta_b;
        double pwm_a = target_a + computePID(error_a, pid_state_a_);
        double pwm_b = target_b + computePID(error_b, pid_state_b_);

        // ── MOTOR DRIVE ───────────────────────────────────────────────────────
        if (active_ && linear_ != 0.0) {
            driveMotorA(pwm_a);
            driveMotorB(pwm_b);
        } else {
            stopMotors();
        }

        // ── ODOMETRY ──────────────────────────────────────────────────────────
        double dist_a = (static_cast<double>(delta_a) / PULSES_PER_REV) * WHEEL_CIRCUMFERENCE;
        double dist_b = (static_cast<double>(delta_b) / PULSES_PER_REV) * WHEEL_CIRCUMFERENCE;
        double dist   = (dist_a + dist_b) / 2.0;
        double dtheta = (dist_a - dist_b) / WHEEL_BASE;

        x_     += dist * std::cos(theta_);
        y_     += dist * std::sin(theta_);
        theta_ += dtheta;

        auto odom        = nav_msgs::msg::Odometry();
        odom.header.stamp    = get_clock()->now();
        odom.header.frame_id = "odom";
        odom.child_frame_id  = "base_link";
        odom.pose.pose.position.x  = x_;
        odom.pose.pose.position.y  = y_;
        odom.twist.twist.linear.x  = dist   / DT;
        odom.twist.twist.angular.z = dtheta / DT;
        odom_pub_->publish(odom);

        if (++debug_counter_ % 25 == 0)
            RCLCPP_INFO(get_logger(), "err_a=%.4f err_b=%.4f pwm_a=%.4f pwm_b=%.4f",
                        error_a, error_b, pwm_a, pwm_b);
        
    }
};


// ─── MAIN ─────────────────────────────────────────────────────────────────────
int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MotorController>());
    rclcpp::shutdown();
    return 0;
}